File: sita/genetic_sur/01/osero.cpp
#include "osero.h"

void osero::play(void){
    bool can = true, old_can = true;
    int line, col;

    srand(this -> srand_num);
    void (osero:: * player[2]) (int *, int *) = {
        play_method[this -> bmethod],
        play_method[this -> wmethod]
    };

    this -> printb();

    while((can = check_all()) || old_can){    
        if (can){
            (this->*player[INT(this -> turn)])(&line, &col);
            this -> put(this -> bw, line, col, this -> turn);
            this -> printb();
        }
        this -> turn = !(this -> turn);
        old_can = can;
    }

    this -> count_last();
}

void osero::random(int * line, int * col){
    int x, y;

    do{
        x = rand() % this -> SIZE;
        y = rand() % this -> SIZE;
    } while (!(this -> check(this -> bw, x, y, this -> turn)));

    *line = x, *col = y;
}

void osero::nleast(int * line, int * col){
    int i = 0, j = 0;
    int place_num, min_place_num = 100;
    int line_ans[this -> SIZE << 1], col_ans[this -> SIZE << 1];
    int num = 0;
    BOARD place = 1;
    BOARD board_leaf[2];

    while (place){
        if (check(this -> bw, i, j, this -> turn)){
            board_leaf[0] = this -> bw[0];
            board_leaf[1] = this -> bw[1];
            this -> put(board_leaf, i, j, this -> turn);
            place_num = this -> check_place(board_leaf, 1, !(this -> turn), !(this -> turn));
            if (place_num < min_place_num){
                min_place_num = place_num;
                line_ans[0] = i;
                col_ans[0] = j;
                num = 0;
            }else if (place_num == min_place_num){
                num++;
                line_ans[num] = i;
                col_ans[num] = j;
            }
        }
        i++;
        if (i == 8) i = 0, j++;
        place = place << 1;
    }

    if (num){
        int place = rand() % (num + 1);
        line_ans[0] = line_ans[place];
        col_ans[0] = col_ans[place];
    }

    *line = line_ans[0];
    *col = col_ans[0];
}

void osero::nmost(int * line, int * col){
    int i = 0, j = 0;
    int place_num, max_place_num = -1;
    int line_ans[this -> SIZE << 1], col_ans[this -> SIZE << 1];
    int num = 0;
    BOARD place = 1;
    BOARD board_leaf[2];

    while (place){
        if (check(this -> bw, i, j, this -> turn)){
            board_leaf[0] = this -> bw[0];
            board_leaf[1] = this -> bw[1];
            this -> put(board_leaf, i, j, turn);
            place_num = this -> check_place(board_leaf, 1, !(this -> turn), this -> turn);
            if (place_num > max_place_num){
                max_place_num = place_num;
                line_ans[0] = i;
                col_ans[0] = j;
                num = 0;
            }else if (place_num == max_place_num){
                num++;
                line_ans[num] = i;
                col_ans[num] = j;
            }
        }
        i++;
        if (i == 8) i = 0, j++;
        place = place << 1;
    }

    if (num) {
        int place = rand() % (num + 1);
        line_ans[0] = line_ans[place];
        col_ans[0] = col_ans[place];
    }

    *line = line_ans[0];
    *col = col_ans[0];
}

int osero::check_place(BOARD * now, int num, bool turn, bool tar_turn){
    int i = 0, j = 0;
    int put_num = 0;
    BOARD place = 1;

    if (turn == tar_turn){
        if (num == this -> read_goal[INT(this -> turn)]){
            while (place){
                if (this -> check(now, i, j, turn)){
                    put_num++;
                }
                i++;
                if (i == 8) i = 0, j++;
                place = place << 1;
            }

            return put_num;
        }else{
            int place_sum = 0;
            BOARD board_leaf[2];

            while (place){
                if (this -> check(now, i, j, turn)){
                    board_leaf[0] = now[0];
                    board_leaf[1] = now[1];
                    put(board_leaf, i, j, turn);
                    place_sum += this -> check_place(board_leaf, num + 1, !turn, tar_turn);
                    put_num++;
                }
                i++;
                if (i == 8) i = 0, j++;
                place = place << 1;
            }

            if (put_num)
                return place_sum / put_num;
            else
                return 0;
        }
    }else{
        int place_sum;
        BOARD board_leaf[2];

        while (place){
            if (this -> check(now, i, j, turn)){
                board_leaf[0] = now[0];
                board_leaf[1] = now[1];
                put(now, i, j, turn);
                place_sum = this -> check_place(board_leaf, num, !turn, tar_turn);
            }
            i++;
            if (i == 8) i = 0, j++;
            place = place << 1;
        }

        return place_sum;
    }
}

void osero::nhand(int * line, int * col){
    int i = 0, j = 0;
    int score = 0, max_score = -100;
    int line_ans[this -> SIZE << 1], col_ans[this -> SIZE << 1];
    int num = 0;
    BOARD place = 1;
    BOARD board_leaf[2];

    while (place){
        if (this -> check(this -> bw, i, j, this -> turn)){
            board_leaf[0] = this -> bw[0];
            board_leaf[1] = this -> bw[1];
            put(board_leaf, i, j, this -> turn);
            score = this -> board_add(board_leaf, 1, !(this -> turn), false);
            if (score == -100){
                int my, opp;
                my = popcount(board_leaf[INT(this -> turn)]);
                opp = popcount(board_leaf[INT(!(this -> turn))]);
                score = my - opp;
            }
            if (score > max_score){
                max_score = score;
                line_ans[0] = i;
                col_ans[0] = j;
                num = 0;
            }else if (score == max_score){
                num++;
                line_ans[num] = i;
                col_ans[num] = j;
            }
        }
        i++;
        if (i == 8) i = 0, j++;
        place = place << 1;
    }

    if (num){
        int place = rand() % (num + 1);
        line_ans[0] = line_ans[place];
        col_ans[0] = col_ans[place];
    }

    *line = line_ans[0];
    *col = col_ans[0];
}

void osero::nhand_custom(int * line, int * col){
    int i = 0, j = 0;
    double score = 0.0, max_score = -100;
    int line_ans[this -> SIZE << 1], col_ans[this -> SIZE << 1];
    int num = 0;
    BOARD place = 1;
    BOARD board_leaf[2];

    while (place){
        if (this -> check(this -> bw, i, j, this -> turn)){
            board_leaf[0] = this -> bw[0];
            board_leaf[1] = this -> bw[1];
            put(board_leaf, i, j, this -> turn);
            score = this -> board_add(board_leaf, 1, !(this -> turn), true);
            if (score == -100.0) score = count(board_leaf);
            if (score > max_score){
                max_score = score;
                line_ans[0] = i;
                col_ans[0] = j;
                num = 0;
            }else if (score == max_score){
                num++;
                line_ans[num] = i;
                col_ans[num] = j;
            }
        }
        i++;
        if (i == 8) i = 0, j++;
        place = place << 1;
    }

    if (num) {
        int place = rand() % (num + 1);
        line_ans[0] = line_ans[place];
        col_ans[0] = col_ans[place];
    }

    *line = line_ans[0];
    *col = col_ans[0];
}

double osero::board_add(BOARD * now, int num, bool turn, bool iscustom){
    if (num == this -> read_goal[INT(this -> turn)]){
        if (iscustom){
            return count(now);
        }else{
            int my, opp;
            my = popcount(now[INT(this -> turn)]);
            opp = popcount(now[INT(!(this -> turn))]);
            return my - opp;
        }
    }

    int i = 0, j = 0, put_num = 0;
    double score_ele, score = 1.0;
    BOARD place = 1;
    BOARD board_leaf[2];

    while (place){
        if (check(now, i, j, turn)){
            board_leaf[0] = now[0];
            board_leaf[1] = now[1];
            put(board_leaf, i, j, turn);
            score_ele = this -> board_add(board_leaf, num + 1, !turn, iscustom);
            if (score_ele == -100.0){
                if (iscustom){
                    score_ele = count(now);
                }else{
                    int my, opp;
                    my = popcount(now[INT(this -> turn)]);
                    opp = popcount(now[INT(!(this -> turn))]);
                    score_ele = my - opp;
                }
            }
            score += score_ele;
            put_num++;
        }
        i++;
        if (i == 8) i = 0, j++;
        place = place << 1;
    }

    if (put_num) return score / put_num;
    else return -100.0;
}

void osero::human(int * line, int * col){
    printf("line: ");
    scanf("%d", line);
    printf("col: ");
    scanf("%d", col);

    *line -= 1;
    *col -= 1;

    while (!(this -> check(this -> bw, *line, *col, this -> turn))){
        printf("It cannot that putting that place. Please once choose.\n");
        printf("line: ");
        scanf("%d", line);
        printf("col: ");
        scanf("%d", col);

        *line -= 1;
        *col -= 1;
    }
}

double osero::count(BOARD * now){
    int my = INT(this -> turn);
    int opp = INT(!(this -> turn));
    int i = 0;
    double score = 0;
    double * myeva = this -> eva[my];
    BOARD place = 1;

    while (place){
        if (now[my] & place)
            score += myeva[i];
        else if (now[opp] & place)
            score -= myeva[i];
        i++;
        place = place << 1;
    }

    return score;
}

void osero::count_last(void){
    int black = 0, white = 1;

    black = popcount(this -> bw[black]);
    white = popcount(this -> bw[white]);

    printf("black: %d, white: %d\n", black, white);
    if (black > white){
        printf("black win!\n");
    }else if (black < white){
        printf("white win!\n");
    }else{
        printf("draw!\n");
    }
}

int osero::popcount(BOARD now){
    now = now - ((now >> 1) & 0x5555555555555555);
    now = (now & 0x3333333333333333) + ((now >> 2) & 0x3333333333333333);
    now = (now + (now >> 4)) & 0x0f0f0f0f0f0f0f0f;
    now = now + (now >> 8);
    now = now + (now >> 16);
    now = now + (now >> 32);

    return now & 0x7f;
}

bool osero::check_all(void){
    int i, j;

    for (i = 0; i < SIZE; i++)
        for (j = 0; j < SIZE; j++)
            if (this -> check(this -> bw, i, j, this -> turn)) return true;

    return false;
}

bool osero::check(BOARD * now, int line, int col, bool turn){
    if (now[INT(TURN::black)] & static_cast<BOARD>(1) << ((line << 3) + col)) return false;
    if (now[INT(TURN::white)] & static_cast<BOARD>(1) << ((line << 3) + col)) return false;

    int x, y;
    int line_x, col_y;
    int my = INT(turn),
        opp = INT(!turn);

    for (x = -1; x <= 1; x++){
        for (y = -1; y <= 1; y++){
            if (x != 0 || y != 0){
                line_x = line + x;
                col_y = col + y;
                while ((now[opp] & static_cast<BOARD>(1) << ((line_x << 3) + col_y))
                       && 0 <= x + line_x && x + line_x < this -> SIZE
                       && 0 <= y + col_y && y + col_y < this -> SIZE){
                    if (now[my] & static_cast<BOARD>(1) << ((line_x + x << 3) + col_y + y))
                        return true;
                    line_x += x;
                    col_y += y;
                }
            }
        }
    }

    return false;
}

void osero::put(BOARD * now, int line, int col, bool turn){
    int x, y;
    int line_x, col_y, num;
    bool my = turn, opp = !turn;
    BOARD inver, place;

    now[my] = now[my] | static_cast<BOARD>(1) << ((line << 3) + col); 

    for (x = -1; x <= 1; x++){
        for (y = -1; y <= 1; y++){
            if (x != 0 || y != 0){
                inver = 0;
                line_x = line + x;
                col_y = col + y;
                place = static_cast<BOARD>(1) << ((line_x << 3) + col_y);
                while ((now[opp] & place)
                       && 0 <= x + line_x && x + line_x < this -> SIZE
                       && 0 <= y + col_y && y + col_y < this -> SIZE){
                    inver = inver | place;
                    line_x += x;
                    col_y += y;
                    place = static_cast<BOARD>(1) << ((line_x << 3) + col_y);
                }
                if (now[my] & place){
                    now[my] = now[my] + inver;
                    now[opp] = now[opp] - inver;
                }
            }
        }
    }
}

void osero::printb(void){
    BOARD place = 1;
    int num = 0;

    printf("\n  ");
    for (int i = 0; i < this -> SIZE; i++) printf(" %d ", i + 1);

    printf("\n -------------------------\n");
    while (place){
        if (num % this -> SIZE == 0) printf("%d", (num >> 3) + 1);
        if (this -> bw[INT(TURN::black)] & place) printf("|○");
        else if (this -> bw[INT(TURN::white)] & place) printf("|●");
        else printf("|  ");
        if (num % this -> SIZE == 7) printf("|\n -------------------------\n");
        num++;
        place = place << 1;
    }

    return;
}

void osero::reset(void){
    this -> bw[INT(TURN::black)] = 0x810000000;
    this -> bw[INT(TURN::white)] = 0x1008000000;    
}

osero::osero(){
    return;
}

osero::osero(int player_b, int player_w){
    this -> bw[INT(TURN::black)] = 0x810000000;
    this -> bw[INT(TURN::white)] = 0x1008000000;
    this -> bmethod = player_b;
    this -> wmethod = player_w;

    return;
}

osero::osero(int player_b, double * eva, int player_w){
    this -> bw[INT(TURN::black)] = 0x810000000;
    this -> bw[INT(TURN::white)] = 0x1008000000;
    this -> bmethod = player_b;
    this -> wmethod = player_w;
    for (int i = 0; i < 64; i++)
        this -> eva[0][i] = eva[i];
    this -> player = 1;
    this -> computer = 0;

    return;
}

osero::osero(int player_b, int player_w, double * eva){
    this -> bw[INT(TURN::black)] = 0x810000000;
    this -> bw[INT(TURN::white)] = 0x1008000000;
    this -> bmethod = player_b;
    this -> wmethod = player_w;
    for (int i = 0; i < 64; i++)
        this -> eva[1][i] = eva[i];
    this -> player = 0;
    this -> computer = 1;

    return;
}

osero::osero(int player_b, double * eva1, int player_w, double * eva2){
    this -> bw[INT(TURN::black)] = 0x810000000;
    this -> bw[INT(TURN::white)] = 0x1008000000;
    this -> bmethod = player_b;
    this -> wmethod = player_w;
    for (int i = 0; i < 64; i++){
        this -> eva[0][i] = eva1[i];
        this -> eva[1][i] = eva2[i];
    }
    this -> player = 1;
    this -> computer = 0;

    return;
}

osero:: ~osero(){
    return;
}

File: combinationAI/48/Osero.java
import java.util.Random;
import java.util.function.BiConsumer;
import java.io.*;
import java.util.ArrayList;
import java.util.Set;

public class Osero extends OseroBase {
    protected static Random rand = new Random(0);

    private static InputStreamReader isr = new InputStreamReader(System.in);
    protected static BufferedReader br = new BufferedReader(isr);

    protected ArrayList<BiConsumer<long[], Boolean>> playMethod = new ArrayList<BiConsumer<long[], Boolean>>();
    protected static int[] readGoal = new int[2];
    protected static double[] customScore = {
         1.0, -0.6,  0.6,  0.4,  0.4,  0.6, -0.6,  1.0,
        -0.6, -0.8,  0.0,  0.0,  0.0,  0.0, -0.8, -0.6,
         0.6,  0.0,  0.8,  0.6,  0.6,  0.8,  0.0,  0.6,
         0.4,  0.0,  0.6,  0.0,  0.0,  0.6,  0.0,  0.4,
         0.4,  0.0,  0.6,  0.0,  0.0,  0.6,  0.0,  0.4,
         0.6,  0.0,  0.8,  0.6,  0.6,  0.8,  0.0,  0.6,
        -0.6, -0.8,  0.0,  0.0,  0.0,  0.0, -0.8, -0.6,
         1.0, -0.6,  0.6,  0.4,  0.4,  0.6, -0.6,  1.0
    };

    public static final boolean PRINT = true;
    public static final boolean NOPRINT = false;

    public Osero(){
        this.setup();
    }

    public Osero(ArrayList<BiConsumer<long[], Boolean>> playMethod){
        this.setup();
        this.playMethod = playMethod;
    }

    public Osero(BiConsumer<long[], Boolean> black, BiConsumer<long[], Boolean> white){
        this.setup();
        this.playMethod.add(black);
        this.playMethod.add(white);
    }

    public void play(boolean printMode){
        boolean can = true, oldCan = true;

        if (printMode) this.printBoard();

        while ((can = this.checkAll()) || oldCan){
            if (can){
                this.playMethod.get(this.turn ? 1:0).accept(this.bw, this.turn);
                if (printMode) this.printBoard();
            }

            this.turn = !this.turn;
            oldCan = can;
        }

        this.countLast();
    }

    public void play(){
        boolean can = true, oldCan = true;

        while ((can = this.checkAll()) || oldCan){
            if (can){
                this.playMethod.get(this.turn ? 1:0).accept(this.bw, this.turn);
            }

            this.turn = !this.turn;
            oldCan = can;
        }
    }

    public static void human(long board[], boolean turn){
        int row = 0, col = 0;
        String rowS, colS;

        do {
            try{
                System.out.print("row: ");
                rowS = br.readLine();
                System.out.print("col: ");
                colS = br.readLine();

                row = Integer.parseInt(rowS);
                col = Integer.parseInt(colS);
                row--; col--;
            }catch (Exception e){
                System.out.println("error. once choose.");
                continue;
            }
        }while (!OseroBase.check(row, col, board, turn));

        OseroBase.put(row, col, board, turn);
    }

    public static void random(long board[], boolean turn){
        int row, col;

        do {
            row = Osero.rand.nextInt(OseroBase.SIZE);
            col = Osero.rand.nextInt(OseroBase.SIZE);
        }while (!OseroBase.check(row, col, board, turn));

        OseroBase.put(row, col, board, turn);
    }

    public static void nHand(long board[], boolean turn){
        Osero.exploreAssist(board, turn, Osero::exploreNHand);
    }

    public static void nHandCustom(long board[], boolean turn){
        Osero.exploreAssist(board, turn, Osero::exploreNHandCustom);
    }

    protected static void exploreAssist(long[] board, boolean turn, FourFunction func){
        double maxScore = -100.0;
        double score;
        int[] rowAns = new int[OseroBase.SIZE << 1];
        int[] colAns = new int[OseroBase.SIZE << 1];
        int placeNum = 0;
        long[] boardLeaf = new long[2];

        int row = -1, col = 0;
        for (int place = 0; place < OseroBase.SIZE << OseroBase.SHIFTNUM; place++){
            row++;
            if (row >= OseroBase.SIZE){
                row = 0;
                col++;
            }
            if (!OseroBase.check(row, col, board, turn)) continue;
            boardLeaf[0] = board[0]; boardLeaf[1] = board[1];
            OseroBase.put(row, col, boardLeaf, turn);
            score = func.getScore(
                boardLeaf,
                turn,
                !turn,
                1
            );
            if (score > maxScore){
                maxScore = score;
                placeNum = 0;
                rowAns[0] = row;
                colAns[0] = col;
            }else if (score == maxScore){
                placeNum++;
                rowAns[placeNum] = row;
                colAns[placeNum] = col;
            }
        }

        if (placeNum > 1){
            int place = rand.nextInt(placeNum+1);
            rowAns[0] = rowAns[place];
            colAns[0] = colAns[place];
        }

        OseroBase.put(rowAns[0], colAns[0], board, turn);
    }

    protected static double exploreNHand(long[] board, boolean nowTurn, boolean turn, int num){
        if (num >= Osero.readGoal[(nowTurn ? 1:0)]) return Osero.count(board, nowTurn);

        int score = 0, placeNum = 0;
        int row = -1, col = 0;
        long[] boardLeaf = new long[2];
        for (int place = 0; place < OseroBase.SIZE << OseroBase.SHIFTNUM; place++){
            row++;
            if (row >= OseroBase.SIZE){
                row = 0;
                col++;
            }
            if (!OseroBase.check(row, col, board, turn)) continue;
            placeNum += 1;
            boardLeaf[0] = board[0]; boardLeaf[1] = board[1];
            OseroBase.put(row, col, boardLeaf, turn);
            score += Osero.exploreNHand(
                boardLeaf,
                nowTurn,
                !turn,
                num + 1    
            );
        }

        if (placeNum > 0) return (double)score / placeNum;
        else              return (double)Osero.count(board, turn);
    }

    protected static double exploreNHandCustom(long[] board, boolean nowTurn, boolean turn, int num){
        if (num >= Osero.readGoal[(nowTurn ? 1:0)]) return Osero.countCustom(board, nowTurn);

        double score = 0, placeNum = 0;
        int row = -1, col = 0;
        long[] boardLeaf = new long[2];
        for (int place = 0; place < OseroBase.SIZE << OseroBase.SHIFTNUM; place++){
            row++;
            if (row >= OseroBase.SIZE){
                row = 0;
                col++;
            }
            if (!OseroBase.check(row, col, board, turn)) continue;
            placeNum += 1;
            boardLeaf[0] = board[0]; boardLeaf[1] = board[1];
            OseroBase.put(row, col, boardLeaf, turn);
            score += Osero.exploreNHandCustom(
                boardLeaf,
                nowTurn,
                !turn,
                num + 1    
            );
        }

        if (placeNum > 0) return score / placeNum;
        else              return Osero.countCustom(board, turn);
    }

    protected static int count(long board[], boolean turn){
        int my, opp;
        if (turn){
            my = 1; opp = 0;
        }else{
            my = 0; opp = 1;
        }

        long place = 1;
        int score = 0;
        while (place != 0){
            if      ((board[my] & place) != 0)  score++;
            else if ((board[opp] & place) != 0) score--;
            place = place << 1;
        }

        return score;
    }

    protected static double countCustom(long board[], boolean turn){
        int my, opp;
        if (turn){
            my = 1; opp = 0;
        }else{
            my = 0; opp = 1;
        }

        long place = 1;
        double score = 0;
        int i = 0;
        while (place != 0){
            if      ((board[my] & place) != 0)  score += Osero.customScore[i];
            else if ((board[opp] & place) != 0) score -= Osero.customScore[i];
            place = place << 1;
            i++;
        }

        return score;
    }

    public static void nLeast(long board[], boolean turn){
        Osero.exploreAssist(board, turn, Osero::exploreNLeast);
    }

    public static void nMost(long board[], boolean turn){
        Osero.exploreAssist(board, turn, Osero::exploreNMost);
    }

    public static double exploreNLeast(long board[], boolean nowTurn, boolean turn, int num){
        int row = -1, col = 0;
        int placeNum = 0;
        double score = 0.;
        long[] boardLeaf = new long[2];

        if (num >= Osero.readGoal[(nowTurn ? 1:0)]){
            for (int place = 0; place < OseroBase.SIZE << OseroBase.SHIFTNUM; place++){
                row++;
                if (row >= OseroBase.SIZE){
                    row = 0;
                    col++;
                }
                if (Osero.check(row, col, board, turn)) placeNum++;
            }

            return -(double)placeNum;
        }

        int nextNum;

        if (nowTurn == turn){
            nextNum = num + 1;
        }else{
            nextNum = num;
        }
        for (int place = 0; place < OseroBase.SIZE << OseroBase.SHIFTNUM; place++){
            row++;
            if (row >= OseroBase.SIZE){
                row = 0;
                col++;
            }
            if (!Osero.check(row, col, board, turn)) continue;
            boardLeaf[0] = board[0]; boardLeaf[1] = board[1];
            placeNum++;
            score += Osero.exploreNLeast(
                boardLeaf,
                nowTurn,
                !turn,
                nextNum
            );
        }

        return placeNum > 0 ? -score / placeNum : 0;
    }

    public static double exploreNMost(long board[], boolean nowTurn, boolean turn, int num){
        int row = -1, col = 0;
        int placeNum = 0;
        double score = 0.;
        long[] boardLeaf = new long[2];

        if (nowTurn == turn && num >= Osero.readGoal[(nowTurn ? 1:0)]){
            for (int place = 0; place < OseroBase.SIZE << OseroBase.SHIFTNUM; place++){
                row++;
                if (row >= OseroBase.SIZE){
                    row = 0;
                    col++;
                }
                if (Osero.check(row, col, board, turn)) placeNum++;
            }

            return (double)placeNum;
        }

        int nextNum;
        if (nowTurn != turn){
            nextNum = num + 1;
        }else{
            nextNum = num;
        }
        for (int place = 0; place < OseroBase.SIZE << OseroBase.SHIFTNUM; place++){
            row++;
            if (row >= OseroBase.SIZE){
                row = 0;
                col++;
            }
            if (!Osero.check(row, col, board, turn)) continue;
            boardLeaf[0] = board[0]; boardLeaf[1] = board[1];
            placeNum++;
            score += Osero.exploreNLeast(
                boardLeaf,
                nowTurn,
                !turn,
                nextNum
            );
        }

        return placeNum > 0 ? score / placeNum : 0;
    }

    public void setPlayMethod(ArrayList<BiConsumer<long[], Boolean>> p){
        if (p.size() != 2){
            System.out.println("playMethod size is wrong.");
            System.exit(-1);
        }
        this.playMethod = p;
    }

    public void setPlayMethod(BiConsumer<long[], Boolean> black, BiConsumer<long[], Boolean> white){
        this.playMethod.clear();
        this.playMethod.add(black);
        this.playMethod.add(white);
    }

    public void setReadGoal(int[] r){
        if (r.length != 2){
            System.out.println("readGoal size is wrong.");
            System.exit(-1);
        }
        this.readGoal = r;
    }

    public void setReadGoal(int black, int white){
        this.readGoal[0] = black;
        this.readGoal[1] = white;
    }

    public void setCustomScore(double[] c){
        if (c.length != 64){
            System.out.println("customScore's length is wrong.");
            System.exit(-1);
        }
        this.customScore = c;
    }

    public void setRandom(long seed){
        this.rand = new Random(seed);
    }
}